### PR TITLE
phys2d

### PR DESCRIPTION
**Pull request by 1b15**
# Description

- added a brax implementation of lunar_lander, allowing it to be run on GPU / TPU
- a new heightmap is generated on reset, which takes up some time because a new brax system needs to be generated as well

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

### Gifs
(the gif preview is slow because it doesn't support 100 fps gifs, view it locally for original speed)

demo with strong initial push:
lunar_lander_example

demo with winds:
lunar_lander_with_wind_example

demo with continuous control:
lunar_lander_continuous_example

# Checklist:

- [x] I have run the [`pre-commit` checks](https://pre-commit.com/) with `pre-commit run --all-files` (see `CONTRIBUTING.md` instructions to set it up)
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

Comments to the checklist:
- a warning is thrown for "unsupported collider pair: box, box"

### CODE REVIEW

**Comment by 1b15** [comment]
New tests are failing because they require the latest brax version, which isn't on PyPi yet.

**Comment by 1b15** [comment]
Notes on bipedal walker:
- main parts, i.e. basic physics, observations (including vectorized raycasting for lidar) and rendering are implemented
- I didn't train any agent and couldn't get the given heuristic to work yet, so some parameter tweaks might be necessary
    - most relevant parameters are editable through the constants at the top of the file
    - some (e.g. hull inertia) are in the config string
    - despite my best efforts, I wouldn't exclude the possibility that actions and observations don't fully match their counterparts in box2d (e.g. order of actuators, sign flips, etc.)
- we still use default numpy for einsum and linspace since it's not in the jax utils
- we'll probably have to make the code a bit nicer once we have a template for brax envs
- hardcore mode isn't implemented yet

out

File: gym/envs/phys2d/bipedal_walker.py [hunk 1]
```diff
@@ -0,0 +1,731 @@
...
+
+    def close(self):
+        if self.screen is not None:
+            import pygame
+
+            pygame.display.quit()
+            pygame.quit()
+
+
+if __name__ == "__main__":
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/phys2d/bipedal_walker.py:650]: This is a legacy feature of bipedal walker, we can remove

File: gym/envs/phys2d/bipedal_walker.py [hunk 2]
```diff
@@ -0,0 +1,731 @@
...
+            [
+                transformed_w / 2 * jp.array([-1, 1, 1, -1]),
+                transformed_h / 2 * jp.array([-1, -1, 1, 1]),
+            ]
+        )
+        rot_centered_verts = (R @ centered_verts).T
+        verts = rot_centered_verts + jp.array([x, y]) * SCALE
+        return verts
+
+    def render(self, mode="human"):
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/phys2d/bipedal_walker.py:531]: Personally, as the phys2d are about performance and vectorisation then I don't think we should include the rendering code for now. Possibly when the jax graphics module is release, we could replace with that

File: gym/envs/phys2d/bipedal_walker.py [hunk 3]
```diff
@@ -0,0 +1,731 @@
...
+        return_info: bool = False,
+        options: Optional[dict] = None,
+    ):
+        super().reset(seed=seed)
+        self.renderer.reset()
+        self.prev_shaping = None
+
+        terrain_map = self._generate_terrain(self.hardcore)
+
+        _SYSTEM_CONFIG = """
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/phys2d/bipedal_walker.py:283]: This should be moved to `init` as we don't understand users change the parameters between resets

File: gym/envs/phys2d/bipedal_walker.py [hunk 4]
```diff
@@ -0,0 +1,731 @@
+from typing import Optional
+
+import brax
+import jax
+import numpy as np
+from brax import jumpy as jp
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/phys2d/bipedal_walker.py:6]: Our plan is that jumpy is solely used by the wrappers for easy compatibility between numpy and jax. 
But with the environment, to ensure performance, we should only use jax.numpy for operations.

We will provide jax based spaces

File: gym/envs/phys2d/bipedal_walker.py [hunk 5]
```diff
@@ -0,0 +1,731 @@
...
+        )
+        self.observation_space = spaces.Box(low, high)
+
+        assert render_mode is None or render_mode in self.metadata["render_modes"]
+        self.render_mode = render_mode
+        self.renderer = Renderer(self.render_mode, self._render)
+        self.screen = None
+        self.clock = None
+
+    def _to_2d(self, arr):
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/phys2d/bipedal_walker.py:183]: Could you add some docstrings to explain the purpose of the function, repeat this comment for the other functions

File: gym/envs/phys2d/bipedal_walker.py [hunk 6]
```diff
@@ -0,0 +1,731 @@
...
+
+    def _get_state(self, info):
+        left_leg_ground_contact, right_leg_ground_contact = (
+            info.contact.vel[(LEFT_LEG, RIGHT_LEG), 2] != 0
+        ).astype(int)
+        joint_angle, joint_vel = self.sys.joints[0].angle_vel(self.qp)
+        joint_angle += [0, 1, 0, 1]
+        joint_vel /= jp.array([6, 4, 6, 4])
+
+        state = [
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/phys2d/bipedal_walker.py:229]: Should this not be a jax / numpy array?

File: setup.py [hunk 7]
```diff
@@ -16,6 +16,7 @@
     "atari": ["ale-py~=0.7.5"],
     "accept-rom-license": ["autorom[accept-rom-license]~=0.4.2"],
     "box2d": ["box2d-py==2.3.5", "pygame==2.1.0"],
+    "phys2d": ["brax>=0.0.13", "pyyaml>=6.0"],
```
Comment (pseudo-rnd-thoughts) [inline setup.py:19]: Why is `pyyaml` required

File: tests/envs/test_phys2d_lunar_lander.py [hunk 8]
```diff
@@ -0,0 +1,24 @@
+import pytest
```
Comment (pseudo-rnd-thoughts) [inline tests/envs/test_phys2d_lunar_lander.py:1]: I believe we can remove all of these tests

File: py.Dockerfile [hunk 9]
```diff
@@ -27,6 +27,6 @@ ENV LD_LIBRARY_PATH="$LD_LIBRARY_PATH:/root/.mujoco/mujoco210/bin"
 COPY . /usr/local/gym/
 WORKDIR /usr/local/gym/
 
-RUN if [ "python:${PYTHON_VERSION}" = "python:3.6.15" ] ; then pip install .[box2d,classic_control,toy_text,other] pytest=="7.0.1" --no-cache-dir; else pip install .[testing] --no-cache-dir; fi
+RUN if [ "python:${PYTHON_VERSION}" = "python:3.6.15" ] ; then pip install .[box2d,phys2d,classic_control,toy_text,other] pytest=="7.0.1" --no-cache-dir; else pip install .[testing] --no-cache-dir; fi
```
Comment (pseudo-rnd-thoughts) [inline py.Dockerfile:30]: As jax is 3.7+ then phys2d should not be added here. This is only for python 3.6 testing.

File: gym/envs/phys2d/lunar_lander.py [hunk 10]
```diff
@@ -0,0 +1,730 @@
+import warnings
```
Comment (pseudo-rnd-thoughts) [inline gym/envs/phys2d/lunar_lander.py:1]: Basically all of the same comments for lunar lander, remove heuristics and `__main__`, move protobuf and change to jax.numpy

**Review by pseudo-rnd-thoughts — CHANGES_REQUESTED** [review]
Thanks for the work, from the gifs, it looks great. I have added a couple of comments to help. Our plan is that soon, we will have a jax environment checker to ensure correctness that we can add as a test.

**Pull request closed without merging**